### PR TITLE
2281 Cold chain authentication

### LINKED ISSUE #2281: Cold Chain authentication
### Login
- The server accepts an object with the shape {username: string, password: string}
- The credentials are validated against the store credentials, which we don't have. Suggest that we add a permission type for users of `COLD_CHAIN_API` and manually (for now) create a user and grant this permission. Next step is to take the store credentials during sync and create a user
- If validated, the server creates a JWT token and returns the cookie. This contains the storeID currently - we will be consuming the JWT so can define the shape

### Validation
- If a JWT is presented, this is checked for a valid storeID

### Estimate
2 days

### Moneyworks Jobcode 🧰
OMS:COLC

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2281

# 👩🏻‍💻 What does this PR do? 
 

# 🧪 How has/should this change been tested? 

Using the branch [#13764-cold-chain-permission](https://github.com/openmsupply/msupply/tree/%2313764-cold-chain-permission) of mSupply, create a new user, assign a default store and grant the cold chain API permission:

Screenshot 2023-10-12 at 4 38 49 PM

Then, you can test using the cold chain app (if you have that installed somewhere) or Postman (other API tools are available).

Send a `POST` request to `http://localhost:8000/coldchain/v1/login` (adjust port / server as necessary)

The body of the post should be 

```
{
    "username": "[put your username here]",
    "password": "[put your password here]"
}
```

and set the content type of the request to JSON (or `application/json` if you are doing this manually)

Screenshot 2023-10-12 at 4 48 06 PM

If you have the details correct, you'll get a `200` response with `{ "success": "true" }` as the body.
Any incorrect details, and you'll get a `401` response code, with some information in the body.

The minimum timeout applies as well, so that an invalid username takes the same amount of time as an invalid password.

You'll also get a cookie returned, this will have the JWT. I haven't added the store id to the JWT - it isn't required by cold chain afaik.

Once the user has authenticated, any subsequent login attempts will use the details in the local database for authentication if the central server is unavailable (as per the current login process).

## 💌 Any notes for the reviewer?
Thanks

## 📃 Documentation

_No user facing changes_

### CODE REVIEW

File: server/server/src/cold_chain.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        Some(user) => {
+            let store_id = match user.default_store() {
+                Some(store) => Some(store.store_row.id.clone()),
+                None => None,
+            };
+            store_id
+        }
+        None => {
+            return Err(AuthError::Denied(AuthDeniedKind::NotAuthenticated(
+                "No default store".to_string(),
```
Comment (andreievg) [inline server/server/src/cold_chain.rs:119]: this is user not found error, I think the default store error should be up one and read something like "No default store that is active on current site", and the `None` return for user would be internal error

**Review by andreievg — COMMENTED** [review]
Sorry @mark-prins, i only just had a chance to have a look at [this issue](https://github.com/openmsupply/msupply/issues/13764). I think the idea for store credentials was to associate credentials witha single store, whereas user permission can exist for a user in multiple stores. Or do we go with default store for user as the store that the sensor data will be associated with ?

> I haven't added the store id to the JWT - it isn't required by cold chain afaik.

I think it's required: https://github.com/openmsupply/msupply/blob/7772bb0692213ecc1e235be92617a395acb923d0/Project/Sources/Methods/coldchain_temperatureLogV1.4dm#L37C1-L37C42

and

https://github.com/openmsupply/msupply/blob/7772bb0692213ecc1e235be92617a395acb923d0/Project/Sources/Methods/coldchain_temperatureLogV1.4dm#L77C4-L77C8

I am wondering which store on the site we will use for subsequent post updates for sensor etc..

While testing i got this error as plain text:

```
Denied(
    InsufficientPermission {
        msg: "Store id not specified in request",
        required_permissions: And(
            [
                HasStoreAccess,
                HasDynamicPermission(
                    ColdChainApi,
                ),
            ],
        ),
    },
)
```

It turns out that cold chain permission should exist on the default store for the user and that store must be active on current site (it wasn't for me), after some time i figured this out by [looking at this](https://github.com/openmsupply/open-msupply/blob/1f78946982c305c4c43c70bae9416d8dc9ef70d1/server/service/src/user_account.rs#L159-L163): 

I wonder how would this appear when mSupply cold chain is being set up (sorry haven't checked)? And same for other errors returned from the server.

I noticed that omSupply cookie is a little bit different from mSupply cookie, there is extra `; HttpOnly; Secure; Path=/coldchain/v1`, do you think that will be an issue (I guess axios would handle it the same way in mSupply could chain app ?)

I really like the idea of re-using the same auth for user in omSupply and not needing to sync credentials, but was wondering about the effect of this on user (deployment team) and our support ? There seems to be a lot more variable in the setup:
* Default store for the user will be used as the store that data is associated with
* User default store must be active on current site
* User default store must have omSupply cold chain permission
* Need to create new user for every cold chain app to avoid confusion (which store it will sync cold chain data to) <- this one is a minor, sync store credentials need to be configured anyway
* No way to configure permissions for cold chain app on remote site

Do you think it's worth asking deplyment team or support about these ? 

Saying all that, I am happy to approve the PR, just pending the request for default store error. It looks like we can change to store credentials quite easy atm (and no one is set to use this in omSupply in the near future).

**Comment by mark-prins** [comment]
> I think the idea for store credentials was to associate credentials with a single store, whereas user permission can exist for a user in multiple stores.

Yes - I've used the default store as the store for which the permission applies, since that is 1:1, although yeah, a user can be allocated to multiple stores

> I haven't added the store id to the JWT - it isn't required by cold chain afaik. / I think it's required:

That is the mSupply API that you are looking at. What I'm talking about here is the JWT provided by the omSupply API, and if the oms API doesn't read the storeID from the JWT (which it doesn't) then there's no need for it


> I noticed that omSupply cookie is a little bit different from mSupply cookie, there is extra ; HttpOnly; Secure; Path=/coldchain/v1, do you think that will be an issue (I guess axios would handle it the same way in mSupply could chain app ?)

Turns out, this is an issue. Axios cannot cope with the `Secure` option unfortunately. I had removed

>  There seems to be a lot more variable in the setup:

I didn't think that there was a lot more, and it's a one off setup. The inability to edit store credentials locally is a pain - and having to create multiple users is also not great. Decided that it was ok though, if documented carefully.

File: server/server/src/cold_chain.rs (hunk 1 above)
Comment (mark-prins) [reply server/server/src/cold_chain.rs:119]: good point - thanks, have updated

**Comment by mark-prins** [comment]
I did originally add a store_id claim to the JWT and decided against it in the end. I didn't want to mess with the existing token, and it did require a reasonable amount of change. I didn't want to add `store_id` to the JWT because the authentication is against a user; so you could have a current store which differed from the store_id in the token. So I went with `default_store_id` but since that's a property of the user anyway it felt like some redundancy for an edge case of usage.

It does mean that when I validate the token sent to the cold chain API, it has to lookup a user (which it does anyway in order to validate) and fetch the default store ID. It's only a small extra though 🤷

**Review by andreievg — APPROVED** [review]
Look good @mark-prins, thanks for explanations, for me the conclusion for now is:

> I didn't think that there was a lot more, and it's a one off setup. The inability to edit store credentials locally is a pain - and having to create multiple users is also not great. Decided that it was ok though, if documented carefully.

* Even though a decision was made, I think we should consult support and deployment team, but I think this auth is abstracted enough, we can change if needed.
* Yeah store_id doesn't need to be store in JWT for now, sorry I backed down on this in the last comment, but didn't emphasised it enough.

**Pull request merged**